src/accumulator/proof.rs

```diff
@@ -241,6 +243,291 @@ impl Proof {
+        new_nodes: Vec<(u64, sha256::Hash)>,
+        before_num_leaves: u64,
+        to_destroy: Vec<u64>,
+    ) -> Result<(Proof, Vec<sha256::Hash>), String> {
+        // Combine the hashes with the targets.
+        let orig_targets_with_hash: Vec<(u64, sha256::Hash)> = self
+            .targets
+            .to_owned()
+            .into_iter()
+            .zip(cached_del_hashes.into_iter())
+            .collect();
+
+        // Attach positions to the proof.
+        let proof_pos = get_proof_positions(
+            &self.targets,
+            before_num_leaves,
+            util::tree_rows(before_num_leaves),
+        );
+        let proof_with_pos = proof_pos
+            .clone()
+            .into_iter()
+            .zip(self.hashes.clone())
+            .collect();
+
+        // Remap the positions if we moved up a after the addition row.
+        let targets_after_remap = Proof::maybe_remap(
+            before_num_leaves,
+            adds.len() as u64,
+            orig_targets_with_hash.clone(),
+        );
+        let mut final_targets = targets_after_remap.clone();
+        let mut new_nodes_iter = new_nodes.iter();
+        let proof_with_pos =
+            Proof::maybe_remap(before_num_leaves, adds.len() as u64, proof_with_pos);
+        // remembers is an index telling what newly created UTXO should be cached
+        for remember in remembers {
+            let remember_target: Option<&sha256::Hash> = adds.get(remember as usize);
+            if let Some(remember_target) = remember_target {
+                let node = new_nodes_iter.find(|(_, hash)| *hash == *remember_target);
+                if let Some((pos, hash)) = node {
+                    final_targets.push((*pos, *hash));
+                }
+            }
+        }
+
+        final_targets.sort();
+        // Move up positions that need to be moved up due to the empty roots
+        // being written over.
+        let num_leaves = before_num_leaves + (adds.len() as u64);
+        let (mut new_target_pos, target_hashes): (Vec<_>, Vec<_>) =
+            final_targets.clone().into_iter().unzip();
+        let mut new_proof_pos = proof_pos;
+
+        for node in to_destroy {
+            (new_target_pos, _) =
+                Proof::calc_next_positions(&vec![node], &final_targets, num_leaves, true)?;
+            (new_proof_pos, _) =
+                Proof::calc_next_positions(&vec![node], &proof_with_pos, num_leaves, true)?;
+        }
+        // Add in proof_hashes to the new_nodes as some needed hashes
+        // will be in the proof hashes.
+        let leaves_after_add = ((before_num_leaves as usize) + adds.len()) as u64;
+        // Grab all the new nodes after this add.
+        let needed_proof_positions = util::get_proof_positions(
+            &new_target_pos,
+            leaves_after_add,
+            util::tree_rows(leaves_after_add),
+        );
+        let mut new_proof: Vec<_> = new_proof_pos.into_iter().zip(self.hashes).collect();
```

Comment:
`new_proof_pos` is guaranteed to be sorted. Some of the positions may move around so it's not guaranteed to be in the same order as `self.hashes`.

In order for the hashes to be in the same order as the positions, we need to update the ordering of the hashes when we shuffle the positions around as well.

Reply:
I think it'll be in the same position because `calc_next_positions` doesn't move targets, the sorting happens below with both hashes and targets.

Reply:
Ah that's another bug then. `calc_next_positions` should result in positions being moved around.

I made this test case for `calc_next_positions`. It's currently failing on the latest commit.

```
    #[test]
    fn test_calc_next_positions() {
        use super::Proof;

        #[derive(Clone)]
        struct Test {
            name: &'static str,
            block_targets: Vec<u64>,
            old_positions: Vec<(u64, sha256::Hash)>,
            num_leaves: u64,
            num_adds: u64,
            append_roots: bool,
            expected: (Vec<u64>, Vec<sha256::Hash>),
        }

        let tests = vec![Test {
            name: "One empty root deleted",
            block_targets: vec![26],
            old_positions: vec![
                (
                    1,
                    bitcoin_hashes::sha256::Hash::from_str(
                        "4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a",
                    )
                    .unwrap(),
                ),
                (
                    13,
                    bitcoin_hashes::sha256::Hash::from_str(
                        "9d1e0e2d9459d06523ad13e28a4093c2316baafe7aec5b25f30eba2e113599c4",
                    )
                    .unwrap(),
                ),
                (
                    17,
                    bitcoin_hashes::sha256::Hash::from_str(
                        "9576f4ade6e9bc3a6458b506ce3e4e890df29cb14cb5d3d887672aef55647a2b",
                    )
                    .unwrap(),
                ),
                (
                    25,
                    bitcoin_hashes::sha256::Hash::from_str(
                        "29590a14c1b09384b94a2c0e94bf821ca75b62eacebc47893397ca88e3bbcbd7",
                    )
                    .unwrap(),
                ),
            ],
            num_leaves: 14,
            num_adds: 14,
            append_roots: false,
            expected: (
                vec![1, 17, 21, 25],
                vec![
                    bitcoin_hashes::sha256::Hash::from_str(
                        "4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a",
                    )
                    .unwrap(),
                    bitcoin_hashes::sha256::Hash::from_str(
                        "9576f4ade6e9bc3a6458b506ce3e4e890df29cb14cb5d3d887672aef55647a2b",
                    )
                    .unwrap(),
                    bitcoin_hashes::sha256::Hash::from_str(
                        "9d1e0e2d9459d06523ad13e28a4093c2316baafe7aec5b25f30eba2e113599c4",
                    )
                    .unwrap(),
                    bitcoin_hashes::sha256::Hash::from_str(
                        "29590a14c1b09384b94a2c0e94bf821ca75b62eacebc47893397ca88e3bbcbd7",
                    )
                    .unwrap(),
                ],
            ),
        }];

        for test in tests {
            let res = Proof::calc_next_positions(
                &test.block_targets,
                &test.old_positions,
                test.num_leaves + test.num_adds,
                test.append_roots,
            )
            .unwrap();

            assert_eq!(res, test.expected, "testcase: \"{}\" fail", test.name);
        }
    }
```

Reply:
> Ah that's another bug then. calc_next_positions should result in positions being moved around.  

Ok, it does result in positions being moved around, but it doesn't return then sorted. I made this way because in some places I don't need hashes, only positions. To avoid unzips, `calc_next_positions` returns a tuple of `Vec<u64>` and `Vec<Hash>`. If I have to sort, then I need to hold a `Vec<(u64, Hash)>`, since I can't sort hashes without positions.  
The code is handling this by sorting if needed. Do you think it's worth changing `calc_next_positions` to produce a sorted `Vec<(u64, Hash)>`?

Reply:
> Do you think it's worth changing `calc_next_positions` to produce a sorted `Vec<(u64, Hash)>`?

When I redid the algorithm for how the proof updates were done, I tried to avoid sorting wherever possible but I couldn't think of how to avoid sorting during position updates. So we can't avoid sorting (for now, maybe there is an algorithm that you can do without sorting).

You could internally use `Vec<(u64, Hash)>` and just unzip before returning. If unzip is too expensive when benchmarked, we can look at ways of reducing the cost.

Reply:
Ok, 826f047 sorts inside `calc_next_position`, I added your test, and it's passing.